dataplane/forwarding/fwdport/ports/tap.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ports
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/openconfig/lemming/dataplane/forwarding/fwdaction"
+	"github.com/openconfig/lemming/dataplane/forwarding/fwdport"
+	"github.com/openconfig/lemming/dataplane/forwarding/infra/fwdcontext"
+	"github.com/openconfig/lemming/dataplane/forwarding/infra/fwdobject"
+	"github.com/openconfig/lemming/dataplane/forwarding/infra/fwdpacket"
+	"golang.org/x/sys/unix"
+
+	log "github.com/golang/glog"
+	fwdpb "github.com/openconfig/lemming/proto/forwarding"
+)
+
+func init() {
+	fwdport.Register(fwdpb.PortType_PORT_TYPE_TAP, tapBuilder{})
+}
+
+// tapPort is a ports that receives from and writes a linux network device.
+type tapPort struct {
+	fwdobject.Base
+	input  fwdaction.Actions
+	output fwdaction.Actions
+	ctx    *fwdcontext.Context // Forwarding context containing the port
+	fd     int
+	file   *os.File
+}
+
+func (p *tapPort) String() string {
+	desc := fmt.Sprintf("Type=%v;DeviceName=%v;<Input=%v>;<Output=%v>", fwdpb.PortType_PORT_TYPE_KERNEL, p.fd, p.input, p.output)
+	if state, err := p.State(nil); err == nil {
+		desc += fmt.Sprintf("<State=%v>;", state)
+	}
+	return desc
+}
+
+func (p *tapPort) Cleanup() {
+	p.input.Cleanup()
+	p.output.Cleanup()
+	p.file.Close()
+	p.input = nil
+	p.output = nil
+}
+
+// Update updates the actions of the port.
+func (p *tapPort) Update(upd *fwdpb.PortUpdateDesc) error {
+	var err error
+	defer func() {
+		if err != nil {
+			p.Cleanup()
+		}
+	}()
+	kernelUpd, ok := upd.Port.(*fwdpb.PortUpdateDesc_Kernel)
+	if !ok {
+		return fmt.Errorf("invalid type for port update")
+	}
+
+	// Acquire new actions before releasing the old ones.
+	if p.input, err = fwdaction.NewActions(kernelUpd.Kernel.GetInputs(), p.ctx); err != nil {
+		return fmt.Errorf("ports: input actions for port %v failed, err %v", p, err)
+	}
+	if p.output, err = fwdaction.NewActions(kernelUpd.Kernel.GetOutputs(), p.ctx); err != nil {
+		return fmt.Errorf("ports: output actions for port %v failed, err %v", p, err)
+	}
+	return nil
+}
+
+func (p *tapPort) process() {
+	go func() {
+		buf := make([]byte, 1500) // TODO: MTU
+		for {
+			n, err := p.file.Read(buf)
```

Comment:
I think we should add a comment to explain why we made this implementation choice.

Reply:
added comment on the tabPort type